src/core/iop/memcard.cpp

```diff
@@ -1,6 +1,308 @@
+{
+	return address + length <= oem_memcard_bytes;
+}
+
+// Create a file on disk with all bits initialized to their default state, 1.
+// The file is unusable in this state; it must be formatted by a game or BIOS.
+bool Memcard::new_memcard_file(std::string file_name)
+{
+	std::ofstream file(file_name);
+
+	if (!file.is_open())
+	{
+		printf("[MCD] (WARN) Attempted to open ofstream on empty file %s, but it either cannot be opened or could not be created! Aborting...", file_name);
+		return false;
+	}
+
+	// Write 0xff to entire card
+	for (int i = 0; i < oem_memcard_bytes; i++) {
+		file.put(0xff);
+	}
+
+	file.close();
+
+	return true;
+}
+
+// Open the memcard file_name, mounting its contents to this memcard.
+bool Memcard::open(std::string file_name)
+{
+	std::ifstream file(file_name);
+
+	if (!file.is_open())
+	{
+		printf("[MCD] (WARN) Attempted to open ifstream on memcard file %s, but it either cannot be opened or does not exist! Aborting...", file_name);
+		return false;
+	}	
+
+	file.read(raw_data.data(), oem_memcard_bytes);
+
+	if (file)
+		printf("[MCD] (INFO) Successfully read all bytes into memory!");
+	else
+		printf("[MCD] (WARN) Failed to read all bytes into memory! Only read %d!", file.gcount());
+
+	file.close();
+	this->file_name = file_name;
+	this->formatted = true;
+
+	for (int i = 0; i < format_string.length(); i++) {
+		if (raw_data.at(i) != format_string.at(i)) {
+			this->formatted = false;
+			break;
+		}
+	}
+
+	// Check if backup block 2 (labelled 2 in documentation, but actually the
+	// first on the card physically) is not erased. Page 16352 is the first page
+	// of backup block 2.
+	if (!is_erase_block_erased(16352 * bytes_per_page))
+	{
+		printf("[MCD] (WARN) Backup block 2 was NOT erased when this memcard was opened! The last write op on this memcard may have failed! The emulated PS2 should be restoring the old data soon, if it hasn't already...");
+	}
+
+	return true;
+}
+
+// Read length bytes from memcard starting at address into destination.
+bool Memcard::read(std::vector<uint8_t> destination, uint32_t address, uint32_t length)
+{
+	if (!address_in_bounds(address, length))
+	{
+		printf("[MCD] (WARN) Attempted to read out-of-bounds memcard address %d! Aborting...", address);
+		return false;
+	}
+
+	memcpy(destination.data(), &raw_data[address], length);
+	return true;
+}
+
+// Write length bytes from source starting at address into memcard.
+bool Memcard::write(std::vector<uint8_t> source, uint32_t address)
+{
+	if (!address_in_bounds(address, source.size()))
+	{
+		printf("[MCD] (WARN) Attempted to write out-of-bounds memcard address %d! Aborting...", address);
+		return false;
+	}
+
+	// Check if a bitwise and of source and address bytes would give a result
+	// different from the source byte. If so, that means something else
+	// currently exists here, or the erase block containing address has not been
+	// wiped yet. 
+	for (int i = 0; i < source.size(); i++)
```

Comment:
If I'm not mistaken, doesn't the function `is_erase_block_erased` serve this purpose?